test/distributed/fsdp/test_fsdp_state_dict.py

```diff
@@ -206,51 +224,62 @@ def test_fsdp_state_dict_with_activation_checkpoint(self, checkpoint_wrap):
 
     @skip_if_lt_x_gpu(2)
     def test_state_dict_rank0_offload_save_load_flow(self):
-        # Test taking checkpoint on rank 0 only, and reload
-        # without redundant CPU memories.
-        model = TransformerWithSharedParams(group=dist.distributed_c10d._get_default_group())
-        my_auto_wrap_policy = partial(
+        """Tests saving a model checkpoint only on rank 0 and loading it on
+        rank 0 only with ``sync_module_states=True`` to emulate the workflow to
+        avoid redundant CPU memory usage."""
+        auto_wrap_policy = partial(
             transformer_auto_wrap_policy,
-            transformer_layer_cls={TransformerEncoderLayer, TransformerDecoderLayer}
+            transformer_layer_cls={TransformerEncoderLayer, TransformerDecoderLayer},
         )
-        model = FSDP(model, auto_wrap_policy=my_auto_wrap_policy)
-        ctx = self._get_state_dict_mgr(
-            model, "state_dict", True
+        fsdp_kwargs = {"auto_wrap_policy": auto_wrap_policy}
+        fsdp_model = TransformerWithSharedParams.init(
+            self.process_group,
+            FSDPInitMode.RECURSIVE,
+            CUDAInitMode.CUDA_BEFORE,
+            fsdp_kwargs,
         )
-        with ctx:
-            state_dict = deepcopy(_get_state_dict(model))
-
-        # All ranks initialize non-FSDP model
-        grp = dist.distributed_c10d._get_default_group()
-        model_new = TransformerWithSharedParams(group=grp)
-        for p in model_new.parameters():
-            with torch.no_grad():
-                p.zero_()
-        # Only rank 0 loads the checkpoint
+        # Force model parameters and buffers to be nonzero
+        with FSDP.summon_full_params(fsdp_model):
+            for tensor in itertools.chain(fsdp_model.parameters(), fsdp_model.buffers()):
+                if torch.count_nonzero(tensor) == 0:
+                    with torch.no_grad():
+                        tensor.add_(torch.tensor(1, dtype=tensor.dtype, device=tensor.device))
+        with self._get_state_dict_mgr(fsdp_model, "state_dict", True):
+            state_dict = deepcopy(_get_state_dict(fsdp_model))
+        # Initialize a non-wrapped model on all ranks
+        new_model = TransformerWithSharedParams.init(
+            self.process_group,
+            FSDPInitMode.NO_FSDP,
+            CUDAInitMode.CUDA_BEFORE,
+        )
+        _zero_model(new_model, zero_buffers=True)
+        # Only load the checkpoint on rank 0
         if self.rank == 0:
-            model_new.load_state_dict(state_dict)
-
-        # TransformerWithSharedParams has a buffer of zeros, so can't pass in
```

Comment:
This was actually slightly incorrect. The issue is not just the `"long_buffer"` of zeros. All linear biases are initialized to zero by default. I added logic to force the model parameters and buffers to be nonzero above. This avoids this caveat altogether.